Modules/Core/Common/include/itkIndexRange.h

```diff
@@ -0,0 +1,420 @@
+    {
+      // Implemented just like the corresponding std::rel_ops operator.
+      return rhs < lhs;
+    }
+
+
+    /** Returns (it1 <= it2) for iterators it1 and it2. */
+    friend bool operator<=(const const_iterator& lhs, const const_iterator& rhs) ITK_NOEXCEPT
+    {
+      // Implemented just like the corresponding std::rel_ops operator.
+      return !(rhs < lhs);
+    }
+
+
+    /** Returns (it1 >= it2) for iterators it1 and it2. */
+    friend bool operator>=(const const_iterator& lhs, const const_iterator& rhs) ITK_NOEXCEPT
+    {
+      // Implemented just like the corresponding std::rel_ops operator.
+      return !(lhs < rhs);
+    }
+
+
+  private:
+    friend class IndexRange;
+
+    // Represents an N-dimensional index that is always zero
+    // Aims towards zero runtime overhead.
+    struct ZeroIndex
+    {
+      // The "index" operator.
+      constexpr IndexValueType operator[](unsigned) const
+      {
+        return 0;
+      }
+
+      // Implicitly converts to a default-initialized itk::Index<N>.
+      constexpr operator IndexType() const
+      {
+        return IndexType();
+      }
+    };
+
+
+    // When BeginAtZero is true, use zero as minimum index, otherwise use itk::Index<N>.
+    using MinIndexType = typename std::conditional<VBeginAtZero, ZeroIndex, IndexType>::type;
+
+    // Private constructor, only used by friend class IndexRange.
+    const_iterator(
+      const IndexType& index,
+      const MinIndexType& minIndex,
+      const IndexType& maxIndex) ITK_NOEXCEPT
+      :
+      // Note: Use parentheses instead of curly braces to initialize data members,
+      // to avoid AppleClang 6.0.0.6000056 compilation error, "no viable conversion..."
+    m_Index(index),
+    m_MinIndex(minIndex),
+    m_MaxIndex(maxIndex)
+    {
+    }
+
+
+    // IndexRange::const_iterator data members:
+
+    // Current (N-dimensional) index.
+    IndexType m_Index;
+
+    // Minimum (N-dimensional) index.
+    MinIndexType m_MinIndex;
+
+    // Maximum (N-dimensional) index.
+    IndexType m_MaxIndex;
+  };
+
+  using iterator = const_iterator;
+  using reverse_iterator = std::reverse_iterator<iterator>;
+  using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+  /** Constructs a range of indices for the specified grid size.
+   */
+  explicit IndexRange(const SizeType& size)
+    :
+    // Note: Use parentheses instead of curly braces to initialize data members,
+    // to avoid AppleClang 6.0.0.6000056 compile errors, "no viable conversion..."
+    m_MinIndex(),
+    m_MaxIndex(CalculateMaxIndex({}, size))
+  {
+  }
+
+
+  /** Constructs a range of indices for the specified image region.
+   * \note This function is unavailable when VBeginAtZero is true (in
+   * case there is a substitution failure, and C++ "SFINAE" kicks in).
+   */
+  template <bool VIsSubstitutionFailure = VBeginAtZero,
+    typename TVoid = typename std::enable_if<!VIsSubstitutionFailure>::type>
```

Comment:
Sorry @hjmjohnson  I read your suggestion to modify "ITKKWStyleOverwrite.txt" too late. Already added a 'dummy' template parameter (`TVoid`). Which is not so bad anyway, as it documents the expected type (`void`) for this template argument: `std::enable_if<!VIsSubstitutionFailure>::type` should yield `void`, if anything at all.